R/util_dfc.R

```diff
@@ -0,0 +1,238 @@
+  col_name_compare_list <- util_dfc.compare_colnames(df1, df2)
+  util.print_pre("Shared column names: " %+%
+                   paste(col_name_compare_list$shared_elements, collapse = ", "))
+  util.print_pre("Columns only in first data frame: " %+%
+                   paste(col_name_compare_list$only_in_first, collapse = ", "))
+  util.print_pre("Columns only in second data frame: " %+%
+                   paste(col_name_compare_list$only_in_second, collapse = ", "))
+
+  # Subset to shared columns if needed.
+  if(length(c(col_name_compare_list$only_in_first, col_name_compare_list$only_in_second)) > 0) {
+    # If there are NO matching columns, exit!
+    if(length(col_name_compare_list$shared_elements) == 0) {
+      util.print_pre("Data frames share no common column names! Comparison is impossible. Exiting.")
+      return()
+    }
+    # otherwise, subset to shared columns
+    util.print_pre("Column names don't perfectly match, so only comparing shared columns: " %+%
+                     paste(col_name_compare_list$shared_elements, collapse = ", "))
+    df1 <- df1[, col_name_compare_list$shared_elements]
+    df2 <- df2[, col_name_compare_list$shared_elements]
+  }
+
+  # Sort columns for apples-to-apples comparison.
+  df1 <- df1[, sort(names(df1))]
+  df2 <- df2[, sort(names(df2))]
+
+
+  ##### COMPARING AND FILTERING ROWS (IF THERE ARE IDENTIFIERS)
+
+  if(length(id_cols) > 0) {
+
+    # Sanity-check if id_cols are in both DFs
+    if(any(!id_cols %in% names(df1)) | any(!id_cols %in% names(df2))) {
+      stop("Error - not all id columns provided are in both data frames.")
+    }
+
+    # Setup - create ID column for each data frame, condensing across multiple columns if necessary
+    df1$temp_util_id <- util_dfc.get_condensed_ids(df1, id_cols)
+    df2$temp_util_id <- util_dfc.get_condensed_ids(df2, id_cols)
+
+    # Determine whether there are any duplicates in either data frame, and report that
+    dup_IDs <- FALSE
+    if(any(duplicated(df1$temp_util_id))) {
+      dup_IDs <- TRUE
+      util.print_pre("WARNING - the first data frame has duplicate ID rows (as defined by the id_cols parameter). This makes precise row-matching impossible.")
+    }
+    if(any(duplicated(df2$temp_util_id))) {
+      dup_IDs <- TRUE
+      util.print_pre("WARNING - the second data frame has duplicate ID rows (as defined by the id_cols parameter). This makes precise row-matching impossible.")
+    }
+
+    # Report on the overlap between unique IDs across data frames
+    unique_id_compare_list <- util_dfc.compare_vecs(unique(df1$temp_util_id), unique(df2$temp_util_id))
+    util.print_pre("COMPARING IDENTIFIERS: ")
+    util.print_pre(length(unique_id_compare_list$shared_elements) %+% " shared unique identifiers, such as: " %+%
+                     paste(head(unique_id_compare_list$shared_elements), collapse = ", "))
+    util.print_pre(length(unique_id_compare_list$only_in_first) %+% " unique identifiers only in first data frame.")
```

Comment:
+1

Reply:
I agree with your earlier point, but in this case I'm only printing the head (first 6 elements) of the shared elements. I think it will be useful for debugging identifier problems to have a few examples of identifiers that DO match as well as a few that don't.